cpp/src/arrow/extension/variable_shape_tensor.cc

```diff
@@ -0,0 +1,341 @@
+      if (x.IsNull()) {
+        uniform_shape.emplace_back(std::nullopt);
+      } else {
+        uniform_shape.emplace_back(x.GetInt64());
+      }
+    }
+    if (uniform_shape.size() != ndim) {
+      return Status::Invalid("uniform_shape size must match ndim. Expected: ", ndim,
+                             " Got: ", uniform_shape.size());
+    }
+  }
+
+  return variable_shape_tensor(value_type, static_cast<int32_t>(ndim), permutation,
+                               dim_names, uniform_shape);
+}
+
+std::shared_ptr<Array> VariableShapeTensorType::MakeArray(
+    std::shared_ptr<ArrayData> data) const {
+  DCHECK_EQ(data->type->id(), Type::EXTENSION);
+  DCHECK_EQ("arrow.variable_shape_tensor",
+            internal::checked_cast<const ExtensionType&>(*data->type).extension_name());
+  return std::make_shared<ExtensionArray>(data);
+}
+
+Result<std::shared_ptr<Tensor>> VariableShapeTensorType::MakeTensor(
+    const std::shared_ptr<ExtensionScalar>& scalar) {
+  const auto ext_scalar = internal::checked_pointer_cast<ExtensionScalar>(scalar);
+  const auto tensor_scalar = internal::checked_pointer_cast<StructScalar>(scalar->value);
+  const auto ext_type =
+      internal::checked_pointer_cast<VariableShapeTensorType>(scalar->type);
+
+  ARROW_ASSIGN_OR_RAISE(const auto shape_scalar, tensor_scalar->field(0));
+  ARROW_ASSIGN_OR_RAISE(const auto data_scalar, tensor_scalar->field(1));
+  const auto shape_array = internal::checked_pointer_cast<Int32Array>(
+      internal::checked_pointer_cast<FixedSizeListScalar>(shape_scalar)->value);
+  const auto data_array =
+      internal::checked_pointer_cast<BaseListScalar>(data_scalar)->value;
+
+  const auto value_type =
+      internal::checked_pointer_cast<FixedWidthType>(ext_type->value_type());
+
+  if (!is_fixed_width(*value_type)) {
+    return Status::Invalid("Cannot convert non-fixed-width values to Tensor.");
+  }
+  if (data_array->null_count() > 0) {
+    return Status::Invalid("Cannot convert data with nulls values to Tensor.");
+  }
+
+  auto permutation = ext_type->permutation();
+  if (permutation.empty()) {
+    for (int64_t j = 0; j < static_cast<int64_t>(ext_type->ndim()); ++j) {
+      permutation.emplace_back(j);
+    }
+  }
+
+  std::vector<int64_t> shape;
+  for (int64_t j = 0; j < static_cast<int64_t>(ext_type->ndim()); ++j) {
+    ARROW_ASSIGN_OR_RAISE(const auto size, shape_array->GetScalar(j));
+    auto size_value = internal::checked_pointer_cast<Int32Scalar>(size)->value;
+    if (size_value < 0) {
+      return Status::Invalid("shape must have non-negative values");
+    }
+    shape.push_back(std::move(size_value));
+  }
+  internal::Permute<int64_t>(permutation, &shape);
+
+  std::vector<std::string> dim_names = ext_type->dim_names();
+  if (!dim_names.empty()) {
+    internal::Permute<std::string>(permutation, &dim_names);
+  }
+
+  std::vector<int64_t> strides;
+  ARROW_CHECK_OK(internal::ComputeStrides(value_type, shape, permutation, &strides));
```

Comment:
Why abort instead of propagating the error here?

Reply:
Switched to `ARROW_ASSIGN_OR_RAISE`.